### ISSUE TITLE
Broken datadog provider behaviour with invalid credentials & existing state

### ISSUE THREAD

**Issue by Jonty** [hashicorp/terraform#9869]
The datadog provider exhibits broken behaviour when empty/invalid datadog credentials are provided and existing state exists for the resources.

### Terraform Version
Terraform v0.7.8

### Affected Resource(s)
- datadog

### Terraform Configuration Files
**terraform.tfstate**
```json
{
    "version": 3,
    "terraform_version": "0.7.8",
    "serial": 3,
    "lineage": "adf75c08-d150-4c57-8314-965a264e2259",
    "modules": [
        {
            "path": [
                "root"
            ],
            "outputs": {},
            "resources": {
                "datadog_monitor.woo": {
                    "type": "datadog_monitor",
                    "depends_on": [],
                    "primary": {
                        "id": "1244119",
                        "attributes": {
                            "escalation_message": "Things are really bad!",
                            "id": "1244119",
                            "message": "Things are bad!",
                            "name": "woo",
                            "query": "ignore this",
                            "thresholds.%": "2",
                            "thresholds.critical": "2",
                            "thresholds.warning": "1",
                            "type": "query alert"
                        },
                        "meta": {},
                        "tainted": false
                    },
                    "deposed": [],
                    "provider": ""
                }
            },
            "depends_on": []
        }
    ]
}
```

**HCL for Scenario 1:**
```hcl
provider "datadog" {
    api_key = ""
    app_key = ""
}

resource "datadog_monitor" "woo" {
    name = "woo"
    type = "query alert"
    message = "Things are bad!"
    escalation_message = "Things are much worse!"
    query = "ignore this"

    thresholds {
        warning = "1"
        critical = "2"
    }
}
```

**HCL for Scenario 2:**
```hcl
# Obviously these are invalid credentials
provider "datadog" {
    api_key = "1f8676578f2e572db89ce28ce2a37988"
    app_key = "abc01bf49005bf31f858024200308a28a67b2d13"
}

resource "datadog_monitor" "woo" {
    name = "woo"
    type = "query alert"
    message = "Things are bad!"
    escalation_message = "Things are much worse!"
    query = "ignore this"

    thresholds {
        warning = "1"
        critical = "2"
    }
}
```

### Debug Output
**For Scenario 1:** https://gist.github.com/Jonty/4e85e31be2213b7699da577d3840ad80
**For Scenario 2:** https://gist.github.com/Jonty/8d0f8e645f8c91f983b6949be8d39628

### Expected Behavior
Terraform should have returned a 403 (or similar) immediately for both scenarios.

### Actual Behavior
**Scenario 1:** Credentials are empty strings. Provider fails to refresh state and throws a JSON parsing error because Datadog returns an unexpected HTML response.

**Scenario 2:** Credentials are invalid. Provider hangs for >60s while refreshing state and returns a 403.

### Steps to Reproduce
1. `terraform apply`

**Comment by ojongerius** [comment]
@Jonty nice catch. Turns out Datadog has a credential validation endpoint. I've raised a PR to get this in https://github.com/zorkian/go-datadog-api at https://github.com/zorkian/go-datadog-api/pull/72, which we can then use in the provider code.

**Comment by ojongerius** [comment]
While validating these changes I noticed that Datadog has a bug in their API: if you use a valid API key but an *invalid* APP key their validation endpoint returns a 200 OK with "valid" set to true. I've raise an issue for this and verified its a bug in that particular endpoint, not across their API. It's a known issue that they intend on fixing.

In the meanwhile these changes will already be helpful for most use cases: both keys are invalid, keys are  not set, invalid api key.

**Comment by stack72** [comment]
Closed via #10567
